### PR TITLE
feat: Support S3 batch export encryption

### PR DESCRIPTION
**Pull request by tomasfarias**
## Problem

S3 batch export cannot be configured to support encryption, which is required by users.

## Changes

- [x] Backend S3 support for SSE S3 encryption with either AES256 or AWS KMS.
  - Annoyingly, boto3 doesn't support `None`, so we have to conditionally set `ServerSideEncryption`, and `SSEKMSKeyId`. Not super happy how this turned out, but for only two arguments this should be fine. Happy to take comments.
- [x] Add Frontend support for creating S3 batch exports with encryption.
  - The `kms_key_id` form field is conditionally added and required if `aws:kms` encryption is selected.
  - Not sure if this is the best UX.

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

Testing is hard as our development setup with MinIO is not configured with a KMS. So, what I did was write a new unit test for encryption that checks if AWS credentials are available and that the `S3_TEST_BUCKET` environment variable is set. This allows developers to test encryption locally by:

1. Logging in to AWS.
2. Creating a test bucket.
3. Running test with `S3_TEST_BUCKET=your-test-bucket DEBUG=1 pytest posthog/temporal/tests/batch_exports/test_s3_batch_export.py::test_s3_export_workflow_with_s3_bucket`

### CODE REVIEW

**Review by benjackwhite — APPROVED** [review]
frontend bits look good 👍

**Comment by tomasfarias** [comment]
Rebased on `master` to resolve conflicts.

**Comment by tomasfarias** [comment]
Rebased on `master` _again_ to fix conflicts with snapshots...

File: frontend/src/scenes/batch_exports/BatchExportEditForm.tsx [hunk 1]
```diff
@@ -208,7 +223,21 @@ export function BatchExportsEditForm(props: BatchExportsEditLogicProps): JSX.Ele
...
+                                            <Field name="kms_key_id" label="AWS KMS Key ID" className="flex-1">
+                                                <LemonInput
+                                                    placeholder={
+                                                        isNew
+                                                            ? 'e.g. 1234abcd-12ab-34cd-56ef-1234567890ab'
+                                                            : 'leave unchanged'
+                                                    }
+                                                />
+                                            </Field>
+                                        )}
```
Comment (tiina303) [inline frontend/src/scenes/batch_exports/BatchExportEditForm.tsx:238]:
```suggestion
                                        {
                                        batchExportConfigForm.encryption == 'aws:kms' && (
                                            <Field name="kms_key_id" label="AWS KMS Key ID" className="flex-1">
                                                <LemonInput
                                                    placeholder={
                                                        isNew
                                                            ? 'e.g. 1234abcd-12ab-34cd-56ef-1234567890ab'
                                                            : 'leave unchanged'
                                                    }
                                                />
                                            </Field>
                                        )}
```
why not simply?

File: posthog/temporal/workflows/s3_batch_export.py [hunk 2]
```diff
@@ -119,10 +124,21 @@ def start(self) -> str:
...
+            optional_kwargs["SSEKMSKeyId"] = self.kms_key_id
 
-        return self.upload_id
+        multipart_response = self.s3_client.create_multipart_upload(
+            Bucket=self.bucket_name,
+            Key=self.key,
+            **optional_kwargs,
+        )
+        upload_id: str = multipart_response["UploadId"]
+        self.upload_id = upload_id
```
Comment (tiina303) [inline posthog/temporal/workflows/s3_batch_export.py:139]: why not
```suggestion
        self.upload_id = multipart_response["UploadId"]
```

**Review by tiina303 — APPROVED** [review]

File: posthog/temporal/workflows/s3_batch_export.py (hunk 2 above)
Comment (tomasfarias) [reply posthog/temporal/workflows/s3_batch_export.py:139]: Type checking fails as `self.upload_id` is `str | None` and this function returns `str`, not `None`, so we need the extra `upload_id: str`

File: frontend/src/scenes/batch_exports/BatchExportEditForm.tsx (hunk 1 above)
Comment (tomasfarias) [reply frontend/src/scenes/batch_exports/BatchExportEditForm.tsx:238]: No reason

**Comment by tomasfarias** [comment]
Rebased on `master` for the daily snapshot conflict resolution...

File: posthog/temporal/workflows/s3_batch_export.py (hunk 2 above)
Comment (tiina303) [reply posthog/temporal/workflows/s3_batch_export.py:139]: shouldn't that work just fine - if it's str | None you can assign it str, I'm confused

Comment (tomasfarias) [reply posthog/temporal/workflows/s3_batch_export.py:139]: You can assign it `str` but then you cannot return it, so you need the extra variable. Not the only way to solve this though, you could have some `isinstance` checks, adding an extra variable seemed like the easiest

**Pull request merged**